File: app/src/main/java/com/lunesu/pengchauferry/FerryRepository.kt
package com.lunesu.pengchauferry

import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope

open class FerryRepository(db: DbOpenHelper) {
    private val ferryDao = FerryDao(db)

    open fun getFerries(from: FerryPier, dow: FerryDay): List<Ferry> {
        return ferryDao.query(from, dow)
    }

    open suspend fun refresh() = coroutineScope {
        awaitAll(
            async { ferryDao.save(PengChauToCentralFetcher.fetch(), FerryPier.Central, FerryPier.PengChau) },
            async { ferryDao.save(KaitoFetcher.fetch(), FerryPier.PengChau, FerryPier.DiscoveryBay, FerryPier.TrappistMonastery) },
            async { ferryDao.save(InterIslandsFetcher.fetch(), FerryPier.PengChau, FerryPier.MuiWo, FerryPier.CheungChau, FerryPier.ChiMaWan) }
//            async { ferry.save(KaitoFetcher2.fetch(), FerryPier.MuiWo, FerryPier.DiscoveryBay) },
        )
        Unit
    }

}

File: iOS/PengChauFerry/FerryRepository.swift
//
//  FerryRepository.swift
//  PengChauFerry
//
//  Created by Lionello Lunesu on 7/3/2020.
//  Copyright © 2020 Lionello Lunesu. All rights reserved.
//

import Foundation

open class FerryRepository {
    private let ferryDao = FerryDao()

    open func getFerries(from: FerryPier, dow: FerryDay) -> [Ferry] {
        return ferryDao.query(from: from, dow: dow)
    }

    open func shouldRefresh() -> Bool {
        return false
    }

    open func refresh(completion: @escaping () -> Void) {
        let group = DispatchGroup()

        group.enter()
        JsonFetcher.fetch { ferries, _ in
            if ferries.count > 80 {
                ferryDao.save(result: ferries, piers: .Central, .PengChau, .HeiLingChau, .DiscoveryBay, .TrappistMonastery, .MuiWo, .CheungChau, .ChiMaWan)
            }
            group.leave()
        }

//        group.enter()
//        PengChauToCentralFetcher.fetch { ferries, _ in
//            if ferries.count > 80 {
//                ferryDao.save(result: ferries, piers: .Central, .PengChau, .HeiLingChau)
//            }
//            group.leave()
//        }

//        group.enter()
//        KaitoFetcher.fetch { ferries, _ in
//            if ferries.count > 50 {
//                ferryDao.save(result: ferries, piers: .PengChau, .DiscoveryBay, .TrappistMonastery)
//            }
//            group.leave()
//        }

//        group.enter()
//        InterIslandsFetcher.fetch { ferries, _ in
//            if ferries.count > 10 {
//                ferryDao.save(result: ferries, piers: .PengChau, .MuiWo, .CheungChau, .ChiMaWan)
//            }
//            group.leave()
//        }

        group.notify(queue: .main) {
            completion()
        }
    }

}
